### PR TITLE
Previous Loot Optimisation

### PR DESCRIPTION
**Pull request by adverserath**
* CalculateValue method will now factor in players luck to increase the modifiers on items, instead of every 600 points. This removes the constant rerolling for better loot and just scales the property weight up from the first roll based on luck.
* Players with 0 luck get no benefit, but a player with 60 luck would get a 1% increase, this is a steady increase 10% at 600 luck, 20% at 1800 luck. Meaning higher luck WILL result in higher property weights on items.

* Treasure Chests now send the cartographers mobile into the loot generation, so treasure hunters will get better loot with luck
* In loot generation the percLow value was being ignored, and setting itself to 0. I've set it to not ignore, but reduced the low property to scale against the luck calculation.

* This is the significant as instead of properties always rolling from 0 > 80, or 0>30 they will always roll at (80/3)>80 (at best, becuase the minimum value is multiplied against the perc from the luck calculation). This again means higher luck will also give higher chances to roll against the minimum values

### CODE REVIEW

File: Scripts/Services/TreasureMaps/TreasureMapInfo.cs [hunk 1]
```diff
@@ -1022,7 +1022,7 @@ public static void Fill(Mobile from, TreasureMapChest chest, TreasureMap tMap)
 
                 if (item != null)
                 {
-                    RunicReforging.GenerateRandomItem(item, from is PlayerMobile pm ? pm.RealLuck : from.Luck, min, max, chest.Map);
```
Comment (adverserath) [inline Scripts/Services/TreasureMaps/TreasureMapInfo.cs:1025]: Add player who found the chest to loot generation

File: Scripts/Services/TreasureMaps/TreasureMapChest.cs [hunk 2]
```diff
@@ -238,7 +238,7 @@ public static void Fill(Mobile from, LockableContainer cont, int level, bool isS
                         int min, max;
                         GetRandomItemStat(out min, out max, propsScale);
```
Comment (adverserath) [inline Scripts/Services/TreasureMaps/TreasureMapChest.cs:240]: Add player who found the chest to loot generation (covers SOS too)

File: Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs [hunk 3]
```diff
@@ -941,30 +941,12 @@ public static bool ApplyResistance(Item item, int value, AosElementAttribute att
...
-            if (reforged)
-            {
-                percent = Utility.RandomMinMax(perclow, perchigh);
-            }
-            else
-            {
-                percent = Utility.RandomMinMax(0, perchigh);
-            }
-
+            decimal percent = Utility.RandomMinMax(perclow, perchigh);
```
Comment (adverserath) [inline Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:944]: A minimum value is always calculated on loot, but was ignored. So the low is now being used again here, to stop everything rolling from 0 upwards

File: Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs [hunk 4]
```diff
@@ -941,30 +941,12 @@ public static bool ApplyResistance(Item item, int value, AosElementAttribute att
...
-            if (percent > perchigh) percent = perchigh;
-
-            int scaledBy = Math.Abs(min - max) + 1;
-
-            scaledBy = 10000 / scaledBy;
-
-            percent *= 10000 + scaledBy;
-
-            return min + (max - min) * percent / 1000001;
+            return Convert.ToInt32(Math.Round(((max - min) * percent / 100) + min));
```
Comment (adverserath) [inline Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:949]: Same calculation as previous, but in less code

File: Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs [hunk 5]
```diff
@@ -1872,13 +1854,14 @@ public static bool GenerateRandomItem(Item item, Mobile killer, BaseCreature cre
...
+        /// <param name="finder">if treasuyre chest then finder is populated</param>
         /// <returns></returns>
-        public static bool GenerateRandomItem(Item item, int luck, int minBudget, int maxBudget, Map map)
+        public static bool GenerateRandomItem(Item item, int luck, int minBudget, int maxBudget, Map map, Mobile finder = null)
         {
             if (item is BaseWeapon || item is BaseArmor || item is BaseJewel || item is BaseHat)
             {
                 int budget = Utility.RandomMinMax(minBudget, maxBudget);
-                GenerateRandomItem(item, null, budget, LootPack.GetLuckChance(luck), ReforgedPrefix.None, ReforgedSuffix.None, map);
+                GenerateRandomItem(item, finder, budget, LootPack.GetLuckChance(luck), ReforgedPrefix.None, ReforgedSuffix.None, map);
```
Comment (adverserath) [inline Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs]: allow treasure chests to pass in mobile (the finder)

File: Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs [hunk 6]
```diff
@@ -1938,18 +1921,23 @@ public static void GenerateRandomItem(Item item, Mobile killer, int basebudget,
 
                     int divisor = GetDivisor(basebudget);
 
-                    double perc = 0.0;
-                    double highest = 0.0;
+                    int luckyRandom = 10000;
```
Comment (adverserath) [inline Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:1924]: Main part of the change:
before: the perc is calculated by constantly rerolling and replacing if its a higher value
now: the perc is rolled once, but then rerolled against the previous number for each 600. then any remaining luck is calculated in too
example: with 900 luck
1st roll as normal:                     luckyRandom =  random(1,10000)=3700
2nd roll for 600 luck :                luckyRandom = random(3700,10000)=6000
3rd roll for 300 remaining luck:  6000+ (random(6000,10000)-6000)*0.5  = 7000

File: Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs [hunk 7]
```diff
@@ -1977,12 +1965,13 @@ public static void GenerateRandomItem(Item item, Mobile killer, int basebudget,
...
 
+
                     if (!IsPowerful(budget))
                     {
                         mods = Math.Max(1, GetProperties(5));
 
-                        perchigh = Math.Max(50, Math.Min(500, budget) / mods);
-                        perclow = Math.Max(20, perchigh / 3);
+                        perchigh = Math.Max(50, Math.Min(550, budget) / mods);
+                        perclow = Math.Max(20, Convert.ToInt32(perchigh / 3 * perc));
```
Comment (adverserath) [inline Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:1974]: the perclow was always ignored, so I've scaled it down based on the perc rolled at the start.
Higher luck will give higher minimums, but will always be a maximum of 33% of the perchigh if perc was 1.0, if perc rolled 0% then you would still roll 0 to perchigh as previously. But luck always matters now.

File: Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs [hunk 8]
```diff
@@ -1992,7 +1981,7 @@ public static void GenerateRandomItem(Item item, Mobile killer, int basebudget,
                         mods = Math.Max(minmods, GetProperties(maxmods));
 
                         perchigh = 100;
-                        perclow = Utility.RandomMinMax(50, 70);
+                        perclow = Convert.ToInt32(Utility.RandomMinMax(50, 70) * perc);
```
Comment (adverserath) [inline Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:1985]: Same as above, this 50-70 was always ignored. So its being reduced against the perc rolled at the start. We would see less Legendary artifacts with 1% properties and more higher values. But 1% is still possible on bad real life luck.

File: Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs (hunk 5 above)
Comment (TrueUO) [reply Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs]: Does production UO take into account t-hunters luck? Just wondering.

File: Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs (hunk 4 above)
Comment (TrueUO) [reply Scripts/Services/LootGeneration/RunicReforging/RunicReforging.cs:949]: This is great. Anything that makes it easier to read / less code.

**Pull request closed without merging**